### PR TITLE
feat(redshift): columns require an id attribute (under feature flag)

### LINKED ISSUE #24234: aws-redshift: allow changing of table column names
### Describe the feature

Currently, the construct is based around the name of a table column being fixed, however redshift enables the user to change the name of the table columns, if needed.

### Use Case

If we have an existing column name that is not distinctive enough, further columns added in the future might "clash" with the "meaning" behind the original table column.

### Proposed Solution

Possibly by implementing an extra `id` attribute to the existing `Column` construct. However this would require a possible refactor to the checking conditions that already exist, due to the `name` attribute being used currently. This is definitely **NOT** a critical change, and infra developers could currently alleviate this problem by using the AWS console, and reworking their tables from there.
However, if this is implementable by just a simple addition of an `id` attribute, then this would be a good step in the right direction I believe for the Redshift Construct.

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.56.0

### Environment details (OS name and version, etc.)

Amazon Linux 2

### PR DESCRIPTION
**Pull request by Rizxcviii**
Adds an `id` attribute to retain tables on changing of the column name. This will essentially fire an `ALTER` statement to rename the column, whilst persisting the id, so columns cannot be dropped.

Closes #24234.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Even though this is module isn't stable, I think we need a feature flag on this. The change itself will cause a replacement and I don't want customers to unintentionally lose data due to a change that is meant to prevent that very thing.

**Comment by Rizxcviii** [comment]
> The change itself will cause a replacement and I don't want customers to unintentionally lose data due to a change that is meant to prevent that very thing.

I don't think it will, since I perform a check to see if the `oldColumn.id` is defined. If the attribute is not defined by the developer, then we compare using the current Boolean expression.
For example,
```typescript
column => oldColumn.name !== column.name
```
now becomes
```typescript
column => oldColumn.id ? oldColumn.id !== column.id : oldColumn.name !== column.name
```
This should ensure backwards compatibility, unless I am mistaken.

**Comment by MrArnoldPalmer** [comment]
@Rizxcviii so that part in the custom resource logic seems to be non-breaking but the new `id` prop on column is not optional, so users would have to provide an ID where there was a name before. Will that still not cause replacement and if not then the ID just isn't being handled for users with existing columns at all?

**Comment by Rizxcviii** [comment]
@MrArnoldPalmer 
Are you referring to the initial assignment of the `id` when users are forced to give an id to a column? If so, then I think the logic within the code will handle this correctly? Within the logic, the newly assigned `id` will belong to the `column` object, but `oldColumn.id` will still be undefined, since this was not specified in the previous deployment.

Or are you referring to the `id` "replacing" the `name` prop. The purpose of the `id` prop is to be persistent, so the `id` will never change throughout the initial assignment. My logic is to keep the `id` static, so thereafter whenever the `name` changes, the column name on redshift can be altered, whilst still persisting the existing column.
The `id` prop will take over the responsibility of the `name`s responsibility of being persistent, to maintain column data; but will allow the `name` of a column to be changed, essentially renaming it on Redshift.
Again, if there is another side effect that I am overlooking, then please correct me.

**Comment by Rizxcviii** [comment]
Actually, now that I'm thinking about it, you are right. If within that initial deployment, the user assumes that the `id` is persistent, they may change the `name` attribute, which will effectively drop their column. I will add a feature flag to address the issue @TheRealAmazonKendra @MrArnoldPalmer

**Comment by TheRealAmazonKendra** [comment]
> Actually, now that I'm thinking about it, you are right. If within that initial deployment, the user assumes that the `id` is persistent, they may change the `name` attribute, which will effectively drop their column. I will add a feature flag to address the issue @TheRealAmazonKendra @MrArnoldPalmer

Great, thank you!

File: packages/@aws-cdk/aws-redshift/README.md [hunk 1]
```diff
@@ -166,7 +166,7 @@ with the supplied schema.
 
 ```ts fixture=cluster
 new Table(this, 'Table', {
-  tableColumns: [{ name: 'col1', dataType: 'varchar(4)' }, { name: 'col2', dataType: 'float' }],
+  tableColumns: [{ id: 'col1', name: 'col1', dataType: 'varchar(4)' }, { id: 'col2', name: 'col2', dataType: 'float' }],
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-redshift/README.md:169]: why not make `id` optional and have it default to `name`?

Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/README.md:169]: Potentially, that's what I was considering.
However, if we default it to the `name` attribute, wouldn't this just be issue that this PR is trying to tackle? (They change the name attribute, and accidently change the id)
I was thinking of making `name` optional and have it default to `id`?

Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/README.md:169]: Actually, my mistake. Given that the `id` is optional, the above error would not happen. However, if the `name` attribute is not the original version, the column will be dropped. Could this be a desired implementation, from a developers perspective?

Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/README.md:169]: This has been done, however I should note that the developer will now have to deploy twice if they decide to "re-use" the id system. We would have to let them know in the feature flag, if you would still like this to included. The below is the check behind this

```typescript
oldColumn.id ? oldColumn.id === column.id ?? column.name : oldColumn.name === column.name
```

Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/README.md:169]: This has been implemented

File: packages/@aws-cdk/aws-redshift/lib/table.ts [hunk 2]
```diff
@@ -297,6 +312,18 @@ export class Table extends TableBase {
...
   }
+
+  private validateColumnIds(columns: Column[]): void {
+    const columnIds = new Set<string>();
+    for (const column of columns) {
+      if (column.id) {
+        if (columnIds.has(column.id)) {
+          throw new Error(`Column id '${column.id}' is not unique.`);
+        }
+        columnIds.add(column.id);
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-redshift/lib/table.ts:323]: `validate` sounds like an operation with no side effects, but this line performs a side effect. Can we change the method name, something like `addColumns()` or `addColumnIds()`?

**Review by comcalvi — CHANGES_REQUESTED** [review]
Just one minor comment, but this looks very good!

File: packages/@aws-cdk/aws-redshift/lib/table.ts (hunk 2 above)
Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/lib/table.ts:323]: This has been done

File: packages/@aws-cdk/aws-redshift/test/integ.database.ts [hunk 3]
```diff
@@ -2,11 +2,16 @@
...
 import * as cdk from '@aws-cdk/core';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { REDSHIFT_COLUMN_ID } from '@aws-cdk/cx-api';
 import * as integ from '@aws-cdk/integ-tests';
 import * as constructs from 'constructs';
 import * as redshift from '../lib';
 
-const app = new cdk.App();
+const useColumnIds = { [REDSHIFT_COLUMN_ID]: false };
+const app = new cdk.App({
```
Comment (MrArnoldPalmer) [inline packages/@aws-cdk/aws-redshift/test/integ.database.ts:12]: I think we should leave this test with the flag disabled and not passing ids so we have both cases in separate tests.

**Review by MrArnoldPalmer — CHANGES_REQUESTED** [review]
Just a test change.

File: packages/@aws-cdk/aws-redshift/test/integ.database.ts (hunk 3 above)
Comment (Rizxcviii) [reply packages/@aws-cdk/aws-redshift/test/integ.database.ts:12]: This is done

**Review by MrArnoldPalmer — DISMISSED** [review]
LGTM! @comcalvi added dnm for you to check out again if needed.

**Review by comcalvi — APPROVED** [review]
lgtm! Nice work!

**Pull request merged**